### PR TITLE
Add subtuple helper to remove layers of Concat, fix bug in loop inv motion

### PR DESCRIPTION
**Pull request by oflatt**

### CODE REVIEW

**Comment by oflatt** [comment]
Looks like there are more bad substitutions to track down!

**Comment by oflatt** [comment]
Fixed another substitution bug, and also fixed the snapshot tests to have "snapshot" in the name for easier testing

File: dag_in_context/src/optimizations/loop_strength_reduction.egg [hunk 1]
```diff
@@ -70,37 +70,38 @@
...
-            (Bop
-                (Mul)
-                (Subst (TmpCtx) (Arg new-arg-ty (TmpCtx)) c-out)
-                (Get (Arg new-arg-ty (TmpCtx)) i)))
+            (Bop (Mul) new-c (Get replace-arg i)))
         (union (Get (Arg new-arg-ty (TmpCtx)) n) new-mul)
 
         ; Subsume the multiplication in the new loop to prevent
         ; from firing loop strength reduction again on the new loop
+        ; subsumption needs something to subsume, add it here
```
Comment (ajpal) [inline dag_in_context/src/optimizations/loop_strength_reduction.egg:99]: in the comment, link to the egglog issue

File: dag_in_context/src/utility/util.egg [hunk 2]
```diff
@@ -68,7 +68,8 @@
 (ruleset subsume-after-helpers)
 ;; After running the `saturating` ruleset, these if statements can be subsumed
 (relation ToSubsumeIf (Expr Expr Expr Expr))
-(rule ((ToSubsumeIf a b c d))
+(rule ((ToSubsumeIf a b c d)
+       (If a b c d))
```
Comment (ajpal) [inline dag_in_context/src/utility/util.egg:72]: link to the egglog issue here too

**Review by ajpal — APPROVED** [review]
two little comments, but lgtm!
thanks for walking through it with me

File: dag_in_context/src/schedule.rs [hunk 3]
```diff
@@ -15,26 +14,33 @@ pub(crate) fn helpers() -> String {
...
+        apply-drop-unions
+        cleanup-drop
 
-    (saturate subst)
-    apply-subst-unions
-    cleanup-subst
-)
+        (saturate subst)
+        apply-subst-unions
+        cleanup-subst)
```
Comment (yihozhang) [inline dag_in_context/src/schedule.rs:30]: `subst` generates `AddContext` that needs to be resolved by the `context` ruleset

Comment (oflatt) [reply dag_in_context/src/schedule.rs:32]: Good catch!

**Pull request merged**